### PR TITLE
Update lib with Bevy ecs 0.5 package

### PR DESCRIPTION
**Pull request by phuoc-thanh**
Hi,

I just update lib with newest bevy ecs package. Please review this, thank you.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -47,15 +47,8 @@ pub struct NetworkingPlugin {
...
-            .resources()
-            .get::<IoTaskPool>()
-            .expect("IoTaskPool resource not found")
-            .0
-            .clone();
-
-        app.add_resource(NetworkResource::new(
-            task_pool,
+        app.insert_resource(NetworkResource::new(
+            TaskPool::new(),
```
Comment (smokku) [inline src/lib.rs:51]: Does that mean that Bevy does not support predefined task pools anymore?
I didn't notice this in the 0.5 release notes.

File: examples/simple.rs [hunk 2]
```diff
@@ -75,19 +74,13 @@ fn send_packets(mut net: ResMut<NetworkResource>, time: Res<Time>, args: Res<Arg
...
-
 fn handle_packets(
     mut net: ResMut<NetworkResource>,
     time: Res<Time>,
-    mut state: ResMut<NetworkReader>,
     network_events: Res<Events<NetworkEvent>>,
 ) {
-    for event in state.network_events.iter(&network_events) {
+    let mut reader = network_events.get_reader();
+    for event in reader.iter(&network_events) {
```
Comment (smokku) [inline examples/simple.rs:83]:
```suggestion
    for event in network_events.get_reader().iter(&network_events) {
```

File: examples/channels.rs [hunk 3]
```diff
@@ -112,7 +112,7 @@ fn server_setup(mut net: ResMut<NetworkResource>) {
 fn client_setup(commands: &mut Commands, mut net: ResMut<NetworkResource>) {
     let mut camera = Camera2dBundle::default();
     camera.orthographic_projection.window_origin = WindowOrigin::BottomLeft;
-    commands.spawn(camera);
+    commands.spawn().insert(camera);
```
Comment (smokku) [inline examples/channels.rs:115]: Is this the recommended way of spawning camera in Bevy 0.5?

File: src/lib.rs (hunk 1 above)
Comment (phuoc-thanh) [reply src/lib.rs:51]: It is not mentioned in migration guide, but new App Builder doesn't have Resources type anymore, so I think yes, it's not supported.

File: examples/channels.rs (hunk 3 above)
Comment (phuoc-thanh) [reply examples/channels.rs:115]: It's the new way to spawn in general.

File: src/lib.rs (hunk 1 above)
Comment (smokku) [reply src/lib.rs:51]: `IoTaskPool` still exists.
https://github.com/bevyengine/bevy/blob/b17f8a4bce5551b418654fffb1fe97ff0f9852f0/crates/bevy_asset/src/lib.rs#L83

File: examples/channels.rs (hunk 3 above)
Comment (smokku) [reply examples/channels.rs:115]: Examples suggest `commands.spawn_bundle(OrthographicCameraBundle::new_2d());` - https://github.com/bevyengine/bevy/blob/v0.5.0/examples/2d/sprite.rs#L16

File: examples/simple.rs (hunk 2 above)
Comment (smokku) [reply examples/simple.rs]: Looks like `.get_reader()` is not needed. https://github.com/bevyengine/bevy/blob/v0.5.0/examples/ecs/event.rs#L46

**Review by smokku — CHANGES_REQUESTED** [review]
I appreciate the effort to get things working, but a proper update requires changes to how things are meant to be built under Bevy 0.5.

File: examples/channels.rs (hunk 3 above)
Comment (phuoc-thanh) [reply examples/channels.rs:115]: `spawn().insert()` or `spawn_bundle()`, both work on new Bevy 0.5 engine, but seem like we need to do a rework on channel example if we decide to upgrade with 0.5 version.

Some bundle is removed or changed (Camera2Bundle)  from API.

File: src/lib.rs (hunk 1 above)
Comment (phuoc-thanh) [reply src/lib.rs]: I've found the Resources is now merged into `World`, so I brought back the task pool. Thank you.

File: examples/simple.rs (hunk 2 above)
Comment (phuoc-thanh) [reply examples/simple.rs]: `get_reader()` is needed. There's many ways to iterate through events, we can use a `EventReader<Event>` from Resource, or use the `get_reader()` from events.

File: examples/channels.rs (hunk 3 above)
Comment (mvlabat) [reply examples/channels.rs:115]: Since the camera bundle comprises the whole entity, it makes sense to just use `spawn_bundle`.

Comment (phuoc-thanh) [reply examples/channels.rs]: Agree, I updated it.

**Comment by phuoc-thanh** [comment]
Hi,

Look like we have a dependency conflict when enabling `bevy_wgpu` feature (included in default feature) on channels example. I've tried to include default feature in dev dependencies and it works on my machine but failed with repo CI check.

@smokku . Do you have any suggestion on this? Thank you.

**Comment by billyb2** [comment]
Hey, so I noticed that I can't do WebRTC client connections using your git branch (or presumably the main repo), and I think it's bvecause the `wbindgen` feature is enabled when using WebRTC, regardless of whether WASM is used or not.

**Comment by mvlabat** [comment]
Hey @billyb2, unfortunately, WebRTC clients are not supported by non-WASM clients (that's a `naia-socket` limitation).

**Comment by billyb2** [comment]
Damn, that's really dissapointing. Do you know of any other cross platform networking libraries?

**Comment by smokku** [comment]
It does not really make sense to use WebRTC connections for native clients. Native can use plain UDP, which is way more efficient.

**Comment by billyb2** [comment]
> It does not really make sense to use WebRTC connections for native clients. Native can use plain UDP, which is way more efficient.

True, although it's kind of annoying to have to write two different network implementations.

**Comment by smokku** [comment]
> True, although it's kind of annoying to have to write two different network implementations.

Can you elaborate?
`naia-socket` job is to hide the difference behind the compilation flag.

I might be missing something, but I am building both native and web client from the same codebase, just switching enabled features.

**Comment by billyb2** [comment]
> > True, although it's kind of annoying to have to write two different network implementations.
> 
> Can you elaborate?
> `naia-socket` job is to hide the difference behind the compilation flag.
> 
> I might be missing something, but I am building both native and web client from the same codebase, just switching enabled features.

So if I use the `use-udp` flag for bevy_networking_turbulence, it'll work with WebRTC and UDP?

**Comment by mvlabat** [comment]
@smokku `naia-socket` (and thus `bevy_networking_turbulence`) can't work in both `webrtc` and `udp` mode at the same time, so there's no way to serve both web and desktop client by one server. If desktop client supported `webrtc`, that would another way to solve the problem.

**Comment by smokku** [comment]
> So if I use the `use-udp` flag for bevy_networking_turbulence, it'll work with WebRTC and UDP?

It will work with UDP only.

But you are building your client differently for Native and Web.
You build a native (binary) client using UDP networking and separate build of web (WASM) client using WebRTC - just by switching feature flags, from the same code base.

**Comment by smokku** [comment]
> there's no way to serve both web and desktop clients by one server.

Yes. You need to build two separate servers (from the same codebase) by switching the compilation feature.

**Comment by billyb2** [comment]
Oh I understand now, thank you!

**Comment by billyb2** [comment]
@phuoc-thanh I sent a PR to upgrade to the latest naia version and fix #21

**Comment by smokku** [comment]
LGTM

**Comment by phuoc-thanh** [comment]
@smokku @mvlabat  I added explicit dependencies that help Cargo pick the right version of `wasm-bindgen`, and CI should work. Please rerun workflows.

**Comment by billyb2** [comment]
@phuoc-thanh I rebased my pull

**Comment by smokku** [comment]
> I added explicit dependencies that help Cargo pick the right version of `wasm-bindgen`

We should rather update CI to use latest `wasm-bindgen` than pin ourselves to legacy version potentially forever.

**Comment by billyb2** [comment]
@smokku @phuoc-thanh the latest version of Bevy on the main branch works with the latest wasm-bindgen FYI.

**Comment by phuoc-thanh** [comment]
Yes I know, but personally I prefer to go with current dependencies config and let update it later when bevy release new major version.

@smokku it seems like the dep issue it not relate to CI, just about Cargo is confusing to pick version of `wasm-bindgen`. As we know that newer versions of bevy work with latest `wasm-bindgen`, can we just go ahead, merge then update the channel example in next PR/update?

**Comment by smokku** [comment]
Fixed by #23

**Pull request closed without merging**